### ISSUE TITLE
Built in logrotation

### ISSUE THREAD

**Issue by schack** [foreversd/forever#128]
Hi

It would be really nice with some sort of built-in logfile rotation feature

Regards
Henrik Schack

**Comment by jamesongithub** [comment]
+1

**Comment by giobero** [comment]
+1

**Comment by nleush** [comment]
+1

**Comment by marcocarvalho** [comment]
+1

**Comment by jeffmcmahan** [comment]
+1

**Comment by cvillamor** [comment]
+1

**Comment by clodio** [comment]
+1

**Comment by ChuanyuWang** [comment]
+1 most wanted feature

**Comment by Redsandro** [comment]
Any news?

http://stackoverflow.com/questions/15231968/nodejs-forever-archive-logs#15232276

**Comment by giannicic** [comment]
Since I'm interested in having this functionality too I can try to do it if nobody is already working on it.
With log rotation i mean a functionality that renames the current log file by appending a progressive number (Eg. error.log.1 ) when the file reach a threshold size and creates a new logfile in place of the old one. 
Are you interested in only this basic feature or also to the followings?
- compress the renamed log files
- delete the renamed log files when reaching an end date
- send the renamed log files by e-mail

**Comment by Redsandro** [comment]
@giannicic Winston already supports [logrotating](https://github.com/winstonjs/winston/pull/205) and `forever` uses `winston`. Perhaps it's as easy as reading a logrotate key in the config json and forwarding it to winston.

I like compression and deletion too. For that purpose, I use `winston` streaming to [logrotate-stream](https://www.npmjs.com/package/logrotate-stream) which supports compression and deleting. Perhaps forever can incorporate something like this.

In other words, be careful you don't reinvent the wheel. It's not as simple as it looks, because renaming the file and creating a new one causes forever to lose its file handle and the log will be `undefined` from there on. You also cannot copy the log and truncate the old one outside of winston because the file cursor will not be reset, causing the truncated file to behave weird.

**Comment by giannicic** [comment]
@Redsandro I've used the word "rename" improperly! :)
Looking at the code seems to me that winston is used only for the forever log file. Maybe is not the right place, since is in forver-monitor code, but I'm also interested in rotating the stdout and stderr log files, and the logging of those files don't use winston. 
You are right about using existing tools rathen that starting from scratch with all the issues mentioned by you ( file handles, etc. ), I was not planning to do that. 
So, I'll have a look about using logrotate-stream for the forever log file. I'll also try to make the change for stderr and stdout in forever-monitor.

**Comment by Redsandro** [comment]
@giannicic interesting. I didn't know about the separate `err` and `out` logs. Is that something new(ish)? When doing `forever list`every node has one single log file.

**Comment by giannicic** [comment]
@Redsandro they are the ones defined by the -o and -e options (don't know if they are new or old) and they are optional. In the list command, as per the code, is showed only the forever logFile (the one that use winston) but is possible to change the conf and show other data like those additional log files.

**Comment by giannicic** [comment]
I must correct myself!
The forever logFile is an aggregate of stdout and stderr and doesn't use winston as evinced in lines 461-468 of [https://github.com/foreverjs/forever/blob/master/lib/forever.js](lib/forever.js)
There is also a comment that explains it.

**Comment by jony89** [comment]
+1

**Comment by leonerd** [comment]
Would a really simple first step here be to have a `SIGHUP` handler that just closes and opens the file again? Then at least logrotate can rename the file then send the signal to `forever`, which will then switch to using the new file. That would be a good start.

**Comment by zlodes** [comment]
+1

**Comment by jp1987** [comment]
+1

**Comment by rejhgadellaa** [comment]
+1

**Comment by cuongv** [comment]
+1

**Comment by HariharanManoharan** [comment]
This must be top priority

**Comment by abdulmoizeng** [comment]
+1

**Comment by artakvg** [comment]
+1

**Comment by BerndErnst** [comment]
+1

**Comment by lisbakke** [comment]
+1 :)

**Comment by osehgol** [comment]
+1

**Comment by wave2** [comment]
+1

**Comment by giltsl** [comment]
+1

**Comment by Joniras** [comment]
Due to logging of console output i thought i may just go in between and override console.log (with winston): 

```text
npm install winston winston-daily-rotate-file --save
```
```ts

import * as winston from "winston";
import * as DailyRotate from "winston-daily-rotate-file";

const UniversalLogger = winston.createLogger({
  level: "info",
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.printf(info => {
      const time = new Date(info.timestamp).toLocaleString("en-EN");
      return time + ":  " + info.message;
    })
  ),
  transports: [
    new DailyRotate({
      filename: './log/%DATE%_server.log',
      datePattern: 'DD.MM',
      zippedArchive: false,
      maxFiles: '7d'
    })
  ]
});
console.log = (command: string, ...args: any[]) => {
  UniversalLogger.info(require('util').formatWithOptions({colors: false}, command, ...args));
};

```

This just rotates 7 files on a daily basis and writes console output (as well as a timestamp) to the file.

**Comment by wvoelcker** [comment]
> Due to logging of console output i thought i may just go in between and override console.log (with winston):
> 
> ```
> npm install winston winston-daily-rotate-file --save
> ```
> 
> ```ts
> import * as winston from "winston";
> import * as DailyRotate from "winston-daily-rotate-file";
> 
> const UniversalLogger = winston.createLogger({
>   level: "info",
>   format: winston.format.combine(
>     winston.format.timestamp(),
>     winston.format.printf(info => {
>       const time = new Date(info.timestamp).toLocaleString("en-EN");
>       return time + ":  " + info.message;
>     })
>   ),
>   transports: [
>     new DailyRotate({
>       filename: './log/%DATE%_server.log',
>       datePattern: 'DD.MM',
>       zippedArchive: false,
>       maxFiles: '7d'
>     })
>   ]
> });
> console.log = (command: string, ...args: any[]) => {
>   UniversalLogger.info(require('util').formatWithOptions({colors: false}, command, ...args));
> };
> ```
> 
> This just rotates 7 files on a daily basis and writes console output (as well as a timestamp) to the file.

Thank you - I also added this to morgan to write express.js logs to the same file

`app.use(
	morgan("tiny", {
		stream: { write: (message) => UniversalLogger.info(message.trim()) },
	})
);`
